File: src/main/java/org/cubexmc/metro/physics/TrainPhysicsEngine.java
package org.cubexmc.metro.physics;

import org.cubexmc.metro.train.TrainInstance;
import org.cubexmc.metro.model.Stop;

public interface TrainPhysicsEngine {

    void init(TrainInstance train);

    void onDeparture(TrainInstance train, Stop fromStop);

    void tick(TrainInstance train, double timeFraction, long currentTick);

    void onArrival(TrainInstance train, Stop atStop, long currentTick);

    void cleanup(TrainInstance train);
}

File: src/main/java/org/cubexmc/metro/physics/TrainPhysicsEngine.kt
package org.cubexmc.metro.physics

import org.cubexmc.metro.model.Stop
import org.cubexmc.metro.train.TrainInstance

interface TrainPhysicsEngine {

    fun init(train: TrainInstance)

    fun onDeparture(train: TrainInstance, fromStop: Stop?)

    fun tick(train: TrainInstance, timeFraction: Double, currentTick: Long)

    fun onArrival(train: TrainInstance, atStop: Stop?, currentTick: Long)

    fun cleanup(train: TrainInstance)
}
